include/DistanceCalculator.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+#ifndef DISTANCE_CALCULATOR_H
+#define DISTANCE_CALCULATOR_H
+
+#include <functional>
+#include "Location.h"
+
+class DistanceCalculator {
+public:
+    using DistanceFunction = std::function<double(const Location&, const Location&)>;
```

Comment:
you don't need this class you must use a map, that holds the functions and find by index, you are creating an object for each search !!